In [1]:
import pandas as pd
import yfinance as yf
from pandas.tseries.offsets import MonthEnd

Creating a stock universe:

In [3]:
Tickers = pd.read_html('https://en.wikipedia.org/wiki/List_of_S%26P_500_companies')[0].Symbol
Tickers

0       MMM
1       AOS
2       ABT
3      ABBV
4       ACN
       ... 
498     YUM
499    ZBRA
500     ZBH
501    ZION
502     ZTS
Name: Symbol, Length: 503, dtype: object

In [6]:
df_ = pd.read_html('https://en.wikipedia.org/wiki/List_of_S%26P_500_companies')[1]
df_.Date = pd.to_datetime(df_.Date.Date)

Let's begin our analysis from 1st January, 2020.
So we remove the 'Added' tickers and add the 'Removed' tickers.

In [8]:
df_ = df_[df_.Date.Date >= '2020-01-01']
df_ = df_.dropna(axis = 0)
Tickers = Tickers[~(Tickers.isin(df_.Added.Ticker))]              # Ticker symbols which are not in the 'Added' list

In [9]:
rem_tick = df_.Removed.Ticker
rem_tick = rem_tick.drop(12)
Tickers = Tickers.append(rem_tick)
Tickers.drop_duplicates(inplace = True)

In [13]:
symbols, prices = [], []
for symbol in Tickers:
    df = yf.download(symbol, start = '2020-01-01')['Adj Close']
    if not df.empty:
        prices.append(df)
        symbols.append(symbol)

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%********

In [15]:
df_prices = pd.concat(prices, axis = 1)
df_prices.columns = symbols
df_prices

MMM        AOS         ABT        ABBV         ACN  \
Date                                                                    
2019-12-31  158.148392  45.041458   82.820442   77.053642  201.957550   
2020-01-02  161.357590  45.164375   82.906258   77.932617  201.554733   
2020-01-03  159.968155  44.767281   81.895561   77.192886  201.219040   
2020-01-06  160.120529  45.050919   82.324623   77.802078  199.905045   
2020-01-07  159.475082  44.748375   81.866943   77.358231  195.589096   
...                ...        ...         ...         ...         ...   
2022-12-30  119.919998  57.240002  109.790001  161.610001  266.839996   
2023-01-03  122.470001  58.919998  109.580002  162.380005  270.260010   
2023-01-04  125.150002  60.660000  111.209999  163.690002  269.339996   
2023-01-05  122.959999  59.380001  110.800003  163.490005  262.980011   
2023-01-06  126.720001  61.439999  112.330002  166.550003  269.209991   

                 ATVI        ADM        ADBE         ADP         AAP  ...  \
Date                                                                  ...   
2019-12-31  58.400173  42.957703  329.809998  160.297272  151.050186  ...   
2020-01-02  57.643391  42.735268  334.429993  160.128006  150.342834  ...   
2020-01-03  57.663044  42.651859  331.809998  159.789566  150.352234  ...   
2020-01-06  58.714684  42.318207  333.709991  160.005814  147.881271  ...   
2020-01-07  59.304386  41.808460  333.390015  158.069046  146.127060  ...   
...               ...        ...         ...         ...         ...  ...   
2022-12-30  76.550003  92.849998  336.529999  238.860001  147.029999  ...   
2023-01-03  76.879997  89.650002  336.920013  237.660004  151.539993  ...   
2023-01-04  76.839996  86.389999  341.410004  238.779999  151.889999  ...   
2023-01-05  76.959999  84.230003  328.440002  233.630005  153.160004  ...   
2023-01-06  77.589996  85.949997  332.750000  240.160004  154.020004  ...   

                  FTI       AIV        HRB       COTY        KSS        HOG  \
Date                                                                          
2019-12-31  15.623205  4.962255  20.391325  11.128141  45.663822  35.504475   
2020-01-02  15.543049  4.859454  20.391325  10.930307  44.014725  35.409004   
2020-01-03  15.725222  4.936314  20.373953  10.940200  44.086422  34.635715   
2020-01-06  15.834528  4.967058  20.434748  10.732474  44.866158  34.893478   
2020-01-07  15.455606  4.902689  20.426065  10.603883  44.964748  34.387501   
...               ...       ...        ...        ...        ...        ...   
2022-12-30  12.190000  7.120000  36.509998   8.560000  25.250000  41.599998   
2023-01-03  11.550000  7.100000  35.070000   8.770000  24.549999  41.119999   
2023-01-04  11.590000  7.350000  35.650002   9.040000  26.070000  41.860001   
2023-01-05  11.820000  7.090000  35.700001   8.980000  26.270000  42.330002   
2023-01-06  12.440000  7.340000  36.680000   9.200000  27.240000  43.830002   

                  JWN         HP       CPRI       ARNC  
Date                                                    
2019-12-31  39.059387  38.922241  38.150002        NaN  
2020-01-02  38.276863  38.785172  38.279999        NaN  
2020-01-03  38.057373  39.282082  37.380001        NaN  
2020-01-06  39.107098  40.395859  37.070000        NaN  
2020-01-07  40.070950  39.873238  37.580002        NaN  
...               ...        ...        ...        ...  
2022-12-30  16.139999  49.570000  57.320000  21.160000  
2023-01-03  15.760000  46.150002  57.770000  21.190001  
2023-01-04  16.620001  46.480000  60.070000  21.610001  
2023-01-05  16.620001  46.570000  60.070000  21.629999  
2023-01-06  16.780001  45.799999  61.400002  22.410000  

[761 rows x 483 columns]

In [16]:
(161.357590-158.148392)/158.148392

0.020292321404064523

df_prices has the closing values of multiple stocks across the given timeperiod

Calculating Returns

In [68]:
df_prices.pct_change()*100

MMM       AOS       ABT      ABBV       ACN      ATVI  \
Date                                                                     
2019-12-31       NaN       NaN       NaN       NaN       NaN       NaN   
2020-01-02  2.029251  0.272872  0.103625  1.140741 -0.199464 -1.295844   
2020-01-03 -0.861128 -0.879188 -1.219102 -0.949193 -0.166529  0.034095   
2020-01-06  0.095282  0.633566  0.523932  0.789172 -0.653010  1.823760   
2020-01-07 -0.403110 -0.671559 -0.555927 -0.570454 -2.158976  1.004346   
...              ...       ...       ...       ...       ...       ...   
2022-12-30 -0.539107 -0.728405 -0.471396 -0.584398 -0.573816 -0.273579   
2023-01-03  2.126420  2.935004 -0.191273  0.476458  1.281672  0.431083   
2023-01-04  2.188291  2.953160  1.487495  0.806748 -0.340418 -0.052030   
2023-01-05 -1.749902 -2.110120 -0.368668 -0.122180 -2.361322  0.156172   
2023-01-06  3.057907  3.469177  1.380865  1.871673  2.368994  0.818604   

                 ADM      ADBE       ADP       AAP  ...       FTI       AIV  \
Date                                                ...                       
2019-12-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2020-01-02 -0.517800  1.400805 -0.105566 -0.468279  ... -0.513059 -2.071637   
2020-01-03 -0.195183 -0.783421 -0.211365  0.006293  ...  1.172053  1.581658   
2020-01-06 -0.782287  0.572615  0.135314 -1.643459  ...  0.695102  0.622824   
2020-01-07 -1.204538 -0.095885 -1.210407 -1.186219  ... -2.393014 -1.295951   
...              ...       ...       ...       ...  ...       ...       ...   
2022-12-30 -0.289953 -0.311034 -0.731442  0.492107  ...  1.077937 -0.973577   
2023-01-03 -3.446416  0.115893 -0.502385  3.067398  ... -5.250200 -0.280899   
2023-01-04 -3.636366  1.332658  0.471259  0.230966  ...  0.346320  3.521127   
2023-01-05 -2.500285 -3.798952 -2.156795  0.836134  ...  1.984465 -3.537412   
2023-01-06  2.042020  1.312263  2.795017  0.561505  ...  5.245346  3.526093   

                 HRB      COTY       KSS       HOG       JWN        HP  \
Date                                                                     
2019-12-31       NaN       NaN       NaN       NaN       NaN       NaN   
2020-01-02  0.000000 -1.777781 -3.611387 -0.268886 -2.003431 -0.352212   
2020-01-03 -0.085156  0.090505  0.162902 -2.183887 -0.573408  1.281206   
2020-01-06  0.298357 -1.898736  1.768661  0.744244  2.758258  2.835361   
2020-01-07 -0.042488 -1.198154  0.219702 -1.450085  2.464628 -1.293749   
...              ...       ...       ...       ...       ...       ...   
2022-12-30 -0.571903 -0.925925  0.677831  1.463411  0.310747 -0.141015   
2023-01-03 -3.944121  2.453271 -2.772280 -1.153845 -2.354394 -6.899331   
2023-01-04  1.653840  3.078671  6.191448  1.799615  5.456857  0.715055   
2023-01-05  0.140250 -0.663721  0.767168  1.122793  0.000000  0.193632   
2023-01-06  2.745097  2.449892  3.692422  3.543586  0.962695 -1.653426   

                CPRI      ARNC  
Date                            
2019-12-31       NaN       NaN  
2020-01-02  0.340753       NaN  
2020-01-03 -2.351091       NaN  
2020-01-06 -0.829324       NaN  
2020-01-07  1.375781       NaN  
...              ...       ...  
2022-12-30  0.209788 -0.235734  
2023-01-03  0.785068  0.141780  
2023-01-04  3.981304  1.982067  
2023-01-05  0.000000  0.092543  
2023-01-06  2.214087  3.606106  

[761 rows x 483 columns]

Cummulative Return: (final price - initial price)/initial price * 100


Cummulative return shows the overall return of a stock over the entire period of analysis

In [78]:
((df_prices.pct_change() + 1).prod() - 1)*100

MMM     -19.872722
AOS      36.407659
ABT      35.630781
ABBV    116.148168
ACN      33.300316
           ...    
HOG      23.449247
JWN     -57.039775
HP       17.670497
CPRI     60.943641
ARNC    223.843925
Length: 483, dtype: float64

Monthly Returns

In [82]:
monthly_return = df_prices.pct_change().resample('M').agg(lambda x: (x+1).prod() - 1)

In [83]:
monthly_return

MMM       AOS       ABT      ABBV       ACN      ATVI  \
Date                                                                     
2019-12-31  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2020-01-31 -0.100669 -0.099024  0.007485 -0.072570 -0.021687 -0.015820   
2020-02-29 -0.050854 -0.073553 -0.116020  0.057887 -0.119975 -0.005985   
2020-03-31 -0.085299 -0.043995  0.024406 -0.111072 -0.095963  0.023224   
2020-04-30  0.112885  0.127106  0.171940  0.094972  0.139444  0.078499   
2020-05-31  0.039966  0.120812  0.030731  0.127372  0.088720  0.129452   
2020-06-30 -0.002876 -0.008000 -0.036768  0.059458  0.064974  0.054460   
2020-07-31 -0.035387  0.026268  0.105006 -0.021508  0.050687  0.088669   
2020-08-31  0.093266  0.017242  0.087739  0.009061  0.067399  0.010771   
2020-09-30 -0.017421  0.078211 -0.005846 -0.085413 -0.058100 -0.030771   
2020-10-31 -0.001374 -0.016051 -0.030963 -0.015197 -0.036420 -0.064484   
2020-11-30  0.089174  0.089379  0.029588  0.228907  0.148356  0.049518   
2020-12-31  0.011926 -0.026461  0.011735  0.024574  0.048657  0.168218   
2021-01-31  0.004977 -0.004620  0.133326 -0.032398 -0.070684 -0.019925   
2021-02-28  0.004797  0.093370 -0.030828  0.051327  0.037120  0.050659   
2021-03-31  0.100651  0.138791  0.000501  0.004455  0.101036 -0.027298   
2021-04-30  0.023147  0.006002  0.005682  0.042836  0.052914 -0.014715   
2021-05-31  0.037507  0.049004 -0.028564  0.015247 -0.026934  0.066455   
2021-06-30 -0.021720  0.013930 -0.006172 -0.004947  0.044762 -0.018612   
2021-07-31 -0.003474 -0.020337  0.047534  0.044003  0.080690 -0.123847   
2021-08-31 -0.008656  0.033983  0.044553  0.038521  0.059431 -0.014949   
2021-09-30 -0.099209 -0.160204 -0.065205 -0.106888 -0.049441 -0.060459   
2021-10-31  0.018584  0.201733  0.095299  0.075926  0.124829  0.010337   
2021-11-30 -0.040610  0.081839 -0.024207  0.005320 -0.003874 -0.250544   
2021-12-31  0.044637  0.086022  0.119027  0.174531  0.159905  0.135324   
2022-01-31 -0.065361 -0.106545 -0.091164  0.021496 -0.144864  0.187585   
2022-02-28 -0.095971 -0.102591 -0.053664  0.079480 -0.106228  0.031515   
2022-03-31  0.001547 -0.068387 -0.018737  0.097043  0.067116 -0.017055   
2022-04-30 -0.031300 -0.081369 -0.037239 -0.086248 -0.106680 -0.050688   
2022-05-31  0.045616  0.028923  0.034890  0.003336 -0.006326  0.030159   
2022-06-30 -0.133164 -0.090486 -0.075004  0.039289 -0.069725 -0.000257   
2022-07-31  0.106870  0.162472  0.006201 -0.054242  0.106967  0.026843   
2022-08-31 -0.122968 -0.107792 -0.056872 -0.063062 -0.058121 -0.018261   
2022-09-30 -0.111379 -0.139415 -0.057379 -0.001859 -0.108022 -0.052873   
2022-10-31  0.138371  0.134180  0.027355  0.101927  0.108288 -0.020716   
2022-11-30  0.013157  0.108799  0.087326  0.100956  0.059986  0.015797   
2022-12-31 -0.048027 -0.057623  0.020543  0.002668 -0.113282  0.035159   
2023-01-31  0.056704  0.073375  0.023135  0.030567  0.008882  0.013586   

                 ADM      ADBE       ADP       AAP  ...       FTI       AIV  \
Date                                                ...                       
2019-12-31  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   
2020-01-31 -0.034304  0.064674  0.005220 -0.177385  ... -0.229944  0.020523   
2020-02-29 -0.152230 -0.017144 -0.097147  0.009335  ... -0.101151 -0.085460   
2020-03-31 -0.065604 -0.077886 -0.111248 -0.296258  ... -0.536253 -0.265259   
2020-04-30  0.055713  0.111237  0.073237  0.295649  ...  0.321958  0.071693   
2020-05-31  0.069179  0.093202 -0.001363  0.152262  ... -0.169473 -0.009389   
2020-06-30  0.015009  0.125996  0.022345  0.024189  ... -0.075676  0.020884   
2020-07-31  0.073434  0.020698 -0.107327  0.053984  ...  0.173976  0.031350   
2020-08-31  0.053606  0.155451  0.046498  0.041095  ... -0.041096 -0.061597   
2020-09-30  0.038651 -0.044722  0.009569 -0.016387  ... -0.180520 -0.064113   
2020-10-31 -0.005378 -0.088351  0.132411 -0.040521  ... -0.123613 -0.053974   
2020-11-30  0.084121  0.07016

In [85]:
df_prices.resample('M').last()          # returns the last price of every month for all stocks

MMM        AOS         ABT        ABBV         ACN  \
Date                                                                    
2019-12-31  158.148392  45.041458   82.820435   77.053627  201.957504   
2020-01-31  142.227768  40.581261   83.440323   71.461815  197.577667   
2020-02-29  134.994904  37.596375   73.759552   75.598534  173.873352   
2020-03-31  123.479965  35.942329   75.559731   67.201630  157.187912   
2020-04-30  137.419052  40.510811   88.551476   73.583916  179.106796   
2020-05-31  142.911163  45.404987   91.272736   82.956467  194.997101   
2020-06-30  142.500092  45.041744   87.916840   87.888931  207.666779   
2020-07-31  137.457458  46.224895   97.148643   85.998650  218.192795   
2020-08-31  150.277618  47.021889  105.672325   86.777901  232.898819   
2020-09-30  147.659637  50.699520  105.054527   79.365944  219.367325   
2020-10-31  147.456818  49.885719  101.801765   78.159828  211.377899   
2020-11-30  160.606094  54.344460  104.813881   96.051163  242.737167   
2020-12-31  162.521515  52.906471  106.043900   98.411568  254.548065   
2021-01-31  163.330429  52.662037  120.182350   95.223267  236.555649   
2021-02-28  164.113876  57.579102  116.477386  100.110809  245.336517   
2021-03-31  180.632141  65.570541  116.535736  100.556824  270.124420   
2021-04-30  184.813278  65.964066  117.197861  104.864243  284.417664   
2021-05-31  191.745056  69.196556  113.850197  106.463058  276.757263   
2021-06-30  187.580383  70.160454  113.147469  105.936409  289.145416   
2021-07-31  186.928757  68.733566  118.525795  110.597969  312.476685   
2021-08-31  185.310623  71.069313  123.806450  114.858307  331.047485   
2021-09-30  166.926132  59.683754  115.733597  102.581291  314.680023   
2021-10-31  170.028275  71.723930  126.762863  110.369919  353.961121   
2021-11-30  163.123413  77.593781  123.694359  110.957054  352.589813   
2021-12-31  170.404694  84.268517  138.417297  130.322540  408.970612   
2022-01-31  159.266937  75.290146  125.798630  133.123901  349.725555   
2022-02-28  143.981979  67.566071  119.047798  143.704575  312.575012   
2022-03-31  144.204758  62.945404  116.817245  157.650040  333.553802   
2022-04-30  139.691086  57.823578  112.467072  144.053055  297.970337   
2022-05-31  146.063187  59.496037  116.391037  144.533600  296.085480   
2022-06-30  126.612892  54.112499  107.661217  150.212173  275.441040   
2022-07-31  140.143967  62.904278  108.328857  142.064438  304.904114   
2022-08-31  122.910805  56.123699  102.167931  133.105606  287.182892   
2022-09-30  109.221100  48.299194   96.305595  132.858124  256.160858   
2022-10-31  124.334145  54.779999   98.940002  146.399994  283.899994   
2022-11-30  125.970001  60.740002  107.580002  161.179993  300.929993   
2022-12-31  119.919998  57.240002  109.790001  161.610001  266.839996   
2023-01-31  126.720001  61.439999  112.330002  166.550003  269.209991   

                 ATVI        ADM        ADBE         ADP         AAP  ...  \
Date                                                                  ...   
2019-12-31  58.400169  42.957706  329.809998  160.297256  151.050140  ...   
2020-01-31  57.476307  41.484077  351.140015  161.134003  124.256134  ...   
2020-02-29  57.132317  35.168968  345.119995  145.480316  125.416122  ...   
2020-03-31  58.459145  32.861729  318.239990  129.295959   88.260643  ...   
2020-04-30  63.048149  34.692570  353.640015  138.765182  114.354851  ...   
2020-05-31  71.209885  37.092579  386.600006  138.576035  131.766769  ...   
2020-06-30  75.087944  37.649288  435.309998  141.672546  134.954010  ...   
2020-07-31  81.745934  40.414021  444.320007  126.467201  142.239380  ...   
2020-08-31  82.626411  42.580467  513.390015  132.347610  148.084671  ...   
2020-09-30  80.083908  44.226234  490.429993  133.614059  145.658020  ...   
2020-10-31  74.919769  43.988407  447.100006  151.306030  139.755798  ...   
2020-11-30  78.629639  47.688778  478.470001  166.555420  140.154343  ...   
2020-12-31  91.85659

Yearly Return: 
(price - price_12_months_before) / price_12_months_before - 1

In [90]:
yearly_return = monthly_return.rolling(12).agg(lambda x: (x + 1).prod() - 1)    #.rolling is used to analyse rolling windows
yearly_return.dropna(inplace = True)
yearly_return

MMM       AOS       ABT      ABBV       ACN      ATVI  \
Date                                                                     
2020-11-30  0.015540  0.206543  0.265556  0.246550  0.201922  0.346394   
2020-12-31  0.027652  0.174617  0.280407  0.277183  0.260404  0.572882   
2021-01-31  0.148372  0.297693  0.440339  0.332506  0.197279  0.566322   
2021-02-28  0.215704  0.531507  0.579150  0.324243  0.411007  0.655579   
2021-03-31  0.462846  0.824326  0.542300  0.496345  0.718481  0.573834   
2021-04-30  0.344888  0.628308  0.323500  0.425097  0.587978  0.437808   
2021-05-31  0.341708  0.523986  0.247363  0.283361  0.419289  0.357611   
2021-06-30  0.316353  0.557676  0.286983  0.205344  0.392353  0.263532   
2021-07-31  0.359903  0.486938  0.220046  0.286043  0.432113  0.016881   
2021-08-31  0.233122  0.511409  0.171607  0.323589  0.421422 -0.008994   
2021-09-30  0.130479  0.177206  0.101653  0.292510  0.434489 -0.039349   
2021-10-31  0.153072  0.437765  0.245193  0.412105  0.674542  0.037483   
2021-11-30  0.015674  0.427814  0.180133  0.155187  0.452558 -0.259138   
2021-12-31  0.048505  0.592783  0.305283  0.324260  0.606654 -0.279999   
2022-01-31 -0.024879  0.429685  0.046731  0.398019  0.478407 -0.127554   
2022-02-28 -0.122670  0.173448  0.022068  0.435455  0.274066 -0.143451   
2022-03-31 -0.201666 -0.040035  0.002416  0.567771  0.234815 -0.134431   
2022-04-30 -0.244150 -0.123408 -0.040366  0.373710  0.047651 -0.166034   
2022-05-31 -0.238243 -0.140188  0.022317  0.357594  0.069838 -0.194417   
2022-06-30 -0.325021 -0.228732 -0.048488  0.417947 -0.047396 -0.179350   
2022-07-31 -0.250281 -0.084810 -0.086031  0.284512 -0.024234 -0.038206   
2022-08-31 -0.336731 -0.210296 -0.174777  0.158868 -0.132502 -0.041440   
2022-09-30 -0.345692 -0.190748 -0.167868  0.295150 -0.185964 -0.033701   
2022-10-31 -0.268744 -0.236238 -0.219487  0.326448 -0.197935 -0.063400   
2022-11-30 -0.227763 -0.217205 -0.130276  0.452634 -0.146515  0.269446   
2022-12-31 -0.296264 -0.320743 -0.206819  0.240077 -0.347533  0.157448   
2023-01-31 -0.204355 -0.183957 -0.107065  0.251090 -0.230225 -0.012135   

                 ADM      ADBE       ADP       AAP  ...       FTI       AIV  \
Date                                                ...                       
2020-11-30  0.110133  0.450744  0.039041 -0.072134  ... -0.604240 -0.186395   
2020-12-31  0.124408  0.516388  0.058567 -0.008970  ... -0.552329  0.061725   
2021-01-31  0.155112  0.306516 -0.013150  0.140712  ... -0.338871 -0.095585   
2021-02-28  0.552306  0.331914  0.151950  0.215109  ... -0.239862  0.025556   
2021-03-31  0.673626  0.493747  0.410977  0.978527  ...  0.539418  0.800474   
2021-04-30  0.755792  0.437450  0.304369  0.665787  ...  0.116229  0.898924   
2021-05-31  0.740149  0.305173  0.369226  0.370335  ...  0.560130  0.947306   
2021-06-30  0.561608  0.345340  0.363461  0.454027  ...  0.778246  0.815478   
2021-07-31  0.433652  0.399059  0.612059  0.426089  ...  0.208428  0.825878   
2021-08-31  0.375207  0.292779  0.536097  0.310306  ...  0.157236  1.007231   
2021-09-30  0.324253  0.173909  0.461783  0.378486  ...  0.603854  1.046163   
2021-10-31  0.425262  0.454619  0.449505  0.551080  ...  0.791190  1.393403   
2021-11-30  0.280201  0.399983  0.354332  0.513750  ... -0.082975  0.831397   
2021-12-31  0.373256  0.133848  0.426015  0.546768  ... -0.153566  0.462121   
2022-01-31  0.535996  0.164636  0.272325  0.576576  ... -0.184045  0.531590   
2022-02-28  0.417721  0.017425  0.197124  0.295153  ... -0.166667  0.500000   
2022-03-31  0.619128 -0.041547  0.230050  0.152303  ...  0.003886  0.192182   
2022-04-30  0.450571 -0.221092  0.188787  0.018923  ... -0.064865 -0.092219   
2022-05-31  0.394568 -0.174601  0.158764  0.022328  ... -0.040745 -0.109220   
2022-06-30  0.308172 -0.374940  0.077461 -0.134831  ... -0.256354 -0.046200   
2022-07-31  0.415888 -0.340251  0.171932 -0.063796  ...  0.120499  0.195402   
2022-08-31  0.494564 -0.437336  0.191276 -0.147558  .

The first 11 columns will be empty as there is no data for the previous 12 months. Consider MMM, the price of the stock increased by around 1.55% from 30 November 2021 to 30 November 2022. We will use this 12 month analysis to select our portfolio.

Let's check which stocks were the top performers from 30 November 2021 till 30 November 2022

In [96]:
nov22 = yearly_return.iloc[0]
portfolio = nov22.nlargest(10)
portfolio

ARNC    2.975434
CARR    2.191842
PENN    1.738654
NVDA    1.281283
BBWI    1.168275
AMD     1.020497
PYPL    0.979477
FDX     0.919388
NOW     0.893419
ALB     0.887933
Name: 2020-11-30 00:00:00, dtype: float64

Now let's check how much these stocks returned in the following month

In [102]:
portfolio_return = monthly_return.loc[portfolio.name + MonthEnd(1), portfolio.index]
portfolio_return

ARNC    0.083242
CARR   -0.006097
PENN    0.233857
NVDA   -0.025568
BBWI   -0.041742
AMD    -0.010253
PYPL    0.093779
FDX    -0.092046
NOW     0.029707
ALB     0.087883
Name: 2020-12-31 00:00:00, dtype: float64

In [101]:
print(portfolio_return.mean())

0.03527639279565392


The average return of the portfolio is 3.52%